### ISSUE TITLE
Result builder source break in Swift 5.8

### ISSUE THREAD

**Issue by calda** [apple/swift#63808]
**Description**

The following sample code compiles successfully in Swift 5.7 but fails to compile in Swift 5.8 (the version in Xcode 14.3 beta 1):

```swift
let nonOptionalString = "non optional string"

func makeArray(@ArrayBuilder<String> _ build: () -> [String]) { 
  print(build())
}

makeArray {
  if let string = nonOptionalString {
    string
  }
}

@resultBuilder
public struct ArrayBuilder<Element> {
  public static func buildExpression(_ value: Element) -> [Element] {
    [value]
  }
  
  public static func buildBlock(_ array: [Element]) -> [Element] {
    array
  }

  public static func buildOptional(_ array: [Element]?) -> [Element] {
    array ?? []
  }
}
```

In Swift 5.7 this builds successfully and prints `["non optional string"]`. 

In Swift 5.8 (the version in Xcode 14.3 beta 1), the build fails with the following error:

> error: initializer for conditional binding must have Optional type, not 'String'
>  if let string = nonOptionalString {

This seems like it was probably a bug in Swift 5.7 that got fixed in 5.8, but since this is a source break for some projects this change may need to be reverted until Swift 6?

**Environment**

Xcode 14.3 beta 1

Also tested on SwiftFiddle with nightly-5.8: https://swiftfiddle.com/zxo5qb4c55dephimafrwip4jfy

**Comment by xedin** [comment]
This was a bug in 5.7 that previously compiled and impacted compile time performance. I have outlined this issue in the forums post - https://forums.swift.org/t/improved-result-builder-implementation-in-swift-5-8/63192 "Structural Diagnostics" sub-section.

**Comment by aheze** [comment]
I asked a question on Stack Overflow about this, just closed it.
https://stackoverflow.com/a/75872523/14351818

**Comment by arthuralee** [comment]
This is still a breaking change though, and anyone who updates their Xcode to the now released version (14.3) will run into this. The offending code could be within swift package dependencies too, so it's not a matter of just fixing your own source code

**Comment by xedin** [comment]
Yes, we are aware of the fact that some of these situations are source breaking as outlined in the post. The issue in this particular case is not with the result builder but with the key path expression.
